### PR TITLE
Marshal MicroTime to json and proto at the same precision

### LINKED ISSUE #111928: MicroTime marshalled at different precision for json and proto
### What happened?

Found some error logs when sending series events to apiserver:
```
eventTime: Invalid value: 2022-08-19 16:00:45.305977 +0800 CST: field is immutable
``` 
seems the later eventTime differed from the first one in series events

After investigation, I found the problem was in the marshal.
MicroTime is supposed to work on microsecond level precision, but init with nanosecond level precision, in this case, `MarshalJSON` generates microsecond level precision data and `Marshal` generates nanosecond level precision, so when unmarshalled, the new time is not considered the exactly same as the original time (they are different at nanosecond level)

### What did you expect to happen?

MicroTime marshal should work at same precision for json and proto, so unmarshal can generate the exactly same MicroTime

### How can we reproduce it (as minimally and precisely as possible)?

```go
func TestTimeUnmarshalFromJSONAndProto(t *testing.T) {
	oldTime := NowMicro()
	jsonData, _ := oldTime.MarshalJSON()
	protoData, _ := oldTime.Marshal()

	var timeFromJSON, timeFromProto MicroTime
	timeFromJSON.UnmarshalJSON(jsonData)
	timeFromProto.Unmarshal(protoData)

	if !timeFromJSON.Equal(&timeFromProto) {
		t.Errorf("time got %v, want %v", timeFromJSON, timeFromProto)
	}
}
```

result:
```
--- FAIL: TestTimeUnmarshalFromJSONAndProto (0.00s)
    /home/dev/go/src/k8s.io/kubernetes/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go:268: time got 2022-08-19 16:39:24.584397 +0800 CST, want 2022-08-19 16:39:24.584397474 +0800 CST
FAIL
FAIL	k8s.io/apimachinery/pkg/apis/meta/v1	0.012s
FAIL
```

### Anything else we need to know?

This causes the series events won't be updated on api-server after created

### Kubernetes version

master

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by haoruan**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Truncate the nanosecond to microsecond when marshaling and unmarshaling proto, make sure the same precision for JSON and proto marshaling/unmarshaling.

#### Which issue(s) this PR fixes:

Fixes #111928

#### Does this PR introduce a user-facing change?

```release-note
Protobuf serialization of metav1.MicroTime timestamps (used in `Lease` and `Event` API objects) has been corrected to truncate to microsecond precision, to match the documented behavior and JSON/YAML serialization. Any existing persisted data is truncated to microsecond when read from etcd.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by haoruan** [comment]
@liggitt PTAL :)

**Comment by liggitt** [comment]
/lgtm
/approve

I updated the release note with more details for users.

The fix looks correct to me, and because the truncation happens on both marshal and unmarshal, it handles already-persisted data that incorrectly kept nanosecond-level precision gracefully. I'd like a second set of eyes on this ahead of merge, just to double check my assertion about compatibility. I think we should probably backport this as well.

cc @kubernetes/api-reviewers 
/assign @deads2k

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go [hunk 1]
```diff
@@ -253,3 +253,37 @@ func TestMicroTimeIsZero(t *testing.T) {
...
+
+func TestTimeUnmarshalJSONAndProtoEqual(t *testing.T) {
+	cases := []struct {
+		name   string
+		input  MicroTime
+		result bool
+	}{
+		{"nanosecond level precision", UnixMicro(123, 123123123), true},
+		{"microsecond level precision", UnixMicro(123, 123123000), true},
+	}
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go:265]: This does a good job of ensuring that round-tripping through protobuf and json marshal → unmarshal ends up with identical data

We also need a test that exercises truncation to nanoseconds during proto unmarshaling... you may need to capture the proto bytes we would serialize a nanosecond-precise microtime to prior to this PR, and embed them in a literal `[]byte{...}` that you pass to a test and ensure they result in a microsecond-precise proto object in memory

**Review by liggitt — COMMENTED** [review]
Noted one other bit of test coverage we need

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go (hunk 1 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go:265]: +1 this doesn't test truncation on unmarshall in its current form.

**Comment by liggitt** [comment]
Thinking through all combinations of old/new proto clients, old/new servers, and good/bad existing data, I think we have one remaining gap:

| Client | Server | Persisted Data | Result |
|---|---|---|---|
| Old | Old | Good | ✅  client and server match |
| Old | Old | Bad | ✅ client and server match and round-trip nanoseconds |
| Old | New | Good | ✅ server no-ops, client round-trips with nanoseconds |
| Old | New | Bad | ✅ server truncates to microseconds, client round-trips with nanoseconds |
| New | Old | Good | ✅ server returns microseconds, client round-trips with microseconds |
| New | Old | Bad | 🚫 server returns nanoseconds, client truncates on round-trip, server rejects update because it mutates the event time |
| New | New | Good | ✅ client and server match and round-trip microseconds |
| New | New | Bad | ✅ client and server match, server truncates to microseconds |

Before changing serialization such that proto clients will drop nanoseconds, we need to make sure older servers will tolerate this.

**Comment by liggitt** [comment]
> I'm surprised it hasn't caused major problems.

I think most components have not switched to the new Events API and are still using core/v1 Events, so are not hitting the stricter `ValidateImmutableField(newEvent.EventTime, oldEvent.EventTime,...` check yet

**Comment by liggitt** [comment]
> > I'm surprised it hasn't caused major problems.
> 
> I think most components have not switched to the new Events API and are still using core/v1 Events, so are not hitting the stricter `ValidateImmutableField(newEvent.EventTime, oldEvent.EventTime,...` check yet

cc @dgrisonnet for visibility to this issue that needs resolving prior to switching more components to the new Events API

**Comment by liggitt** [comment]
> Before changing serialization such that proto clients will drop nanoseconds, we need to make sure older servers will tolerate this.

I opened https://github.com/kubernetes/kubernetes/pull/112183 to tolerate this, which we can backport independent of this serialization change.

**Comment by deads2k** [comment]
Other than the test gap noted, this looks correct to me.  I prefer to merge and backport 
https://github.com/kubernetes/kubernetes/pull/112183 before merging this.

**Comment by haoruan** [comment]
@liggitt @lavalamp @deads2k thanks for the review, one test case was added to test truncation on unmarshaling old nanosecond-precise microtime.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go [hunk 2]
```diff
@@ -253,3 +253,68 @@ func TestMicroTimeIsZero(t *testing.T) {
...
+}
+
+func TestMicroTimeProtoUnmarshalRaw(t *testing.T) {
+	cases := []struct {
+		name     string
+		input    []byte
+		expected MicroTime
+	}{
+		// input is generated by Timestamp{123, 123123123}.Marshal()
+		{"nanosecond level precision", []byte{8, 123, 16, 179, 235, 218, 58}, UnixMicro(123, 123123000)},
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time_test.go:302]: lgtm, I verified that this unmarshaled to `UnixMicro(123, 123123123)` when the truncation in Unmarshal is commented out

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/112183 is merged and picked back to 1.22+

/lgtm
/approve
/hold cancel

After this merges to master, go ahead and open picks of this change to 1.23+ and we can figure out the timing we want to merge them to release branches

**Comment by haoruan** [comment]
Thanks, opened picks to 1.23-1.25

**Pull request merged**